File: base/src/main/java/com/zhangteng/base/mvp/base/BasePresenter.java
package com.zhangteng.base.mvp.base;

import java.lang.ref.WeakReference;

public class BasePresenter<V extends BaseView> {
    protected WeakReference<V> view;

    public void attachView(V view) {
        if (!isAttach()) {
            this.view = new WeakReference<>(view);
        }
    }

    public void detachView() {
        if (isAttach()) {
            view.clear();
            view = null;
        }
    }

    boolean isAttach() {
        return view != null &&
                view.get() != null;
    }
}

File: base/src/main/java/com/zhangteng/base/mvp/base/BasePresenter.kt
package com.zhangteng.base.mvp.base

import java.lang.ref.WeakReference

open class BasePresenter<V : IView?> : IPresenter<V> {
    protected var mView: WeakReference<V?>? = null
    protected var mModel: IModel? = null

    override fun attachView(view: V?) {
        if (!isAttach()) {
            this.mView = WeakReference(view)
        }
    }

    override fun detachView() {
        if (isAttach()) {
            mView?.clear()
            mView = null
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        mModel?.onDestroy()
        mModel = null
    }

    override fun isAttach(): Boolean {
        return mView != null &&
                mView!!.get() != null
    }
}